kvserver: 10% synthetic reproposals

tbg opened the pull request:
Hoping to shake out something relevant for
https://github.com/cockroachdb/cockroach/issues/97102 or
https://github.com/cockroachdb/cockroach/issues/97141.

Not intended for merge. Can maybe turn this into a metamorphic var down
the road. For now running manually via `./experiment.sh`.

To date, it hasn't produced anything. For the first runs, I ~immediately
got a closed timestamp regression. This was puzzling, but then also
disappeared, so I think my OSX clock might have been adjusting over
those few minutes.

Epic: none
Release note: None

tbg committed:
kvserver: fix false positive in 'outstanding reproposal' assertion

In #94633, I introduced[^1] an assertion that attempted to catch cases
in which we might otherwise accidentally end up applying a proposal
twice.

This assertion had a false positive, see the updated comment within.

I was able to reproduce the failure within ~minutes via
`./experiment.sh` in #97173 as of 33dcdef.

Better testing of these cases would be desirable. Unfortunately, while
there is an abstraction over command application (`apply.Task`), most
of the logic worth testing lives in `(*replicaAppBatch)` which is
essentially a `*Replica` with more moving parts attached. This does
not lend itself well to unit testing.

I had a run[^1][^2][^3] earlier this year to make log application
standalone, but then didn't have enough time to follow through.
It would be desirable to do so at a later date, perhaps with
the explicit goals of having interactions like the one discussion
in this PR unit become testable.

No release note because unreleased (except perhaps in an alpha).

[3]: #93309
[2]: #93266
[1]: #93239

Closes #94633.

[^1]: https://github.com/cockroachdb/cockroach/pull/94633/files#diff-50e458584d176deae52b20a7c04461b3e4110795c8c9a307cf7ee6696ba6bc60R238

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97173)
<!-- Reviewable:end -->

tbg committed:
scripts: gceworker: don't anchor absolute paths at repo in get

Now one may `./scripts/gceworker.ssh get /home/foo bar`.
Previously it would try to copy from
`go/src/github.com/cockroachdb/cockroach/home/foo/bar`.

Epic: none
Release note: None

tbg committed:
raftlog: don't embed raftpb.Entry in Entry

`raftpb.Entry` is a `fmt.Stringer`, so by embedding it we were making
the string representation of structs embedding it worse.

Concretely, in #97173 I noticed that `*replicatedCmd` ended up just
printing the `raftpb.Entry` that it wrapped, which was not helpful.

Epic: none
Release note: None

tbg committed:
kvserver: improve output on closed timestamp regression

Epic: none
Release note: None

tbg committed:
kvserver: standalone printRaftTail

Epic: none
Release note: None

tbg committed:
kvserver: downgrade and improve a log message

It's not Warning-worthy since this is hit in production, often under
some kind of overload.

And if we're logging this, might as well include enough information
to identify the log entry for which a reproposal was attempted.

This is helpful for investigations such as #70894, #97287, #97102, etc.

Epic: none
Release note: None

tbg committed:
kvserver: 10% error rate in LAI assignment and reproposals

This is an experiment to expose a class of bugs in the replication layer
related to reproposals.

Epic: none
Release note: None

tbg committed:
kvserver: re-submittal log improvements

Epic: none
Release note: None

tbg committed:
Merge branches 'resubmit-log', 'reprop-warning', 'ct-regression-improve', 'avoid-stringer-embedding' and 'gceworker-get-abs' into force-reproposals

tbg committed:
kvpb

tbg committed:
raftentry
